ldms/src/sampler/zfs_zpool/zfs_zpool.c

```diff
@@ -0,0 +1,523 @@
+	if (rc != 0) {
+		log_fn(LDMSD_LERROR,
+		       SAMP " sample():zfs_pool print_stat() failed: %d\n", rc);
+		base_sample_end(sampler_base);
+		goto err1;
+	}
+	/* this is where the rubber meets the pavement */
+
+	base_sample_end(sampler_base);
+
+ err1:
+	return rc;
+}
+
+static void term(struct ldmsd_plugin *self)
+{
+	log_fn(LDMSD_LDEBUG, SAMP " term() called\n");
+	base_set_delete(sampler_base);
+	base_del(sampler_base);
+	sampler_base = NULL;
+}
+
+static ldms_set_t get_set(struct ldmsd_sampler *self)
+{
+	return NULL;
+}
+
+static const char *usage(struct ldmsd_plugin *self)
+{
+	log_fn(LDMSD_LDEBUG, SAMP " usage() called\n");
+	return "config name=" SAMP " " BASE_CONFIG_SYNOPSIS BASE_CONFIG_DESC;
+}
+
+struct ldmsd_plugin *get_plugin(ldmsd_msg_log_f pf)
+{
+	static struct ldmsd_sampler plugin = {
+		.base = {
+			 .name = SAMP,
+			 .type = LDMSD_PLUGIN_SAMPLER,
+			 .term = term,
+			 .config = config,
+			 .usage = usage,
+			 },
+		.get_set = get_set,
+		.sample = sample,
+	};
+
+	log_fn = pf;
+	log_fn(LDMSD_LDEBUG,
+	       SAMP " get_plugin() called (" PACKAGE_STRING ")\n");
+
+	return &plugin.base;
+}
+
+/*
+ * top-level vdev stats are at the pool level moving to its own plugin
+ */
+
+static int get_detailed_pool_stats(nvlist_t * nvroot, const char *pool_name)
+{
+	nvlist_t *nv_ex;
+	uint64_t cap;
+	ldms_mval_t record_instance;
+	uint_t c;
+	vdev_stat_t *vs;
+	int rc = 0;
+
+	if (nvlist_lookup_uint64_array(nvroot, ZPOOL_CONFIG_VDEV_STATS,
+				       (uint64_t **) & vs, &c) != 0) {
+		return (1);
+	}
+
+	if (nvlist_lookup_nvlist(nvroot,
+				 ZPOOL_CONFIG_VDEV_STATS_EX, &nv_ex) != 0) {
+		rc = 6;
+	}
+
+	if (rc == 0) {
+		record_instance = ldms_record_alloc(sampler_base->set,
+						    index_store.vdev_rec_idx);
+
+		if (record_instance == NULL) {
+			log_fn(LDMSD_LDEBUG,
+			       SAMP
+			       ": ldms_record_alloc() failed, resizing metric set\n");
+			resize_metric_set();
+			record_instance = ldms_record_alloc(sampler_base->set,
+							    index_store.
+							    vdev_rec_idx);
+			if (record_instance == NULL)
+				rc = 2;
+		}
+	}
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+					     record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, zpool_metric_ids[0],
+					  pool_name);
+		/* zpool state  1 */
+		ldms_record_array_set_str(record_instance, zpool_metric_ids[1],
+					  zpool_state_to_name((vdev_state_t)
+							      vs->vs_state,
+							      (vdev_aux_t) vs->
+							      vs_aux));
+		/* total     2 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[2],
+				    vs->vs_space);
+		/* allocated 3 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[3],
+				    vs->vs_alloc);
+		/* free     4 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[4],
+				    vs->vs_space - vs->vs_alloc);
+		/* used     5 */
+		cap =
+		    (vs->vs_space ==
+		     0) ? 0 : (vs->vs_alloc * 10000 / vs->vs_space) / 100;
```

Comment:
```suggestion
		                cap = (vs->vs_space == 0) ? 0 : (vs->vs_alloc * 100 / vs->vs_space);
```

No need to divide by 100 here if you only multiple by 100.

Reply:
no more division by 100. to complicated.